[[TreeView Widget]] Mobile Support Enhancements

bwmilby opened the pull request:
General enhancements to the TreeView Widget that support use on the mobile platform and for use as a menu.

* Ability to customize font
* Ability to customize row height
* Ability to trim leading characters from key (to support custom sorting)
* Property/messages necessary to support mobile scroller
* Ability to disable hover
* Save `alternate row backgrounds` property
* Ability to set icon size (height)
* Improve key/value trimming logic (less extra space on right side of widget)
* When `readOnly`, only save space for one icon on right

- [x] Need to add bug fix/enhancement note

bwmilby committed:
[[TreeView Widget]] Adjust row height to reflect font size changes

bwmilby committed:
[[Tree View Widget]]  Add charsToTrimFromKey property

Include a new property that will allow custom key sort display by removing
a set number of characters from each key when rendering the widget.  Array
will need to be configured with characters prepended to each key that will
sort in the desired order.

bwmilby committed:
[Tree View Widget] Add hilitedElementIsFolded property

Use the hilitedElementFoldState code to create a boolean property that can
adjust the fold state of the selected element.

bwmilby committed:
[Tree View Widget] Add scroller support and custom text height

These are new properties, a new handler, and a new message that will be
useful to control the widget view with a `mobileScroller`.

Added `formattedHeight` property
Added `viewTopPosition` property
Added `scroll` property (also includes `vScroll`)

Added `setViewTopPosition` handler

Added `formattedHeightChanged` message

The custom text height property allows for changing the default height
of a row to a custom value instead of using what would be calculated.

Added `textHeight` property

bwmilby committed:
[Tree View Widget] Clear hilitedElement bug fix

If no element was selected, setting the hilitedElement to empty would
throw an error.  Added an additional test to ensure that the widget did
not attempt to unselect and empty path.

Updated description of the textHeight property.

bwmilby committed:
[Tree View Widget] Add vScrollbar property

bwmilby committed:
[Tree View Widget] Add showHover property

Add `showHover` property to allow the hover shading to be disabled.
This is useful on mobile where the hover is not something that makes
sense in that UI.

bwmilby committed:
[Tree View Widget] Save `alternate row backgrounds` property

The `alternate row backgrounds` property was not being saved/restored.
Added the property to the OnSave/OnLoad handlers.

bwmilby committed:
[Tree View Widget] Add iconHeight property

Allow the setting of the size of the icons displayed in the widget.

bwmilby committed:
[Tree View Widget] Handle font changes and improve trimming logic

* When the font size changes, re-evaluate the width of the ellipsis.
* When the font name changes, re-evaluate the width of the ellipsis.
* When in readOnly mode, only save space for one icon on the right.
* If full string displayed, remove ellipsis length from the returned width.
* `displayWidth` was reporting the incorrect width when a string was being
trimmed.  Keep track of the previous width to return a proper value.
* Prevent the last character (key/value) from being turned into an ellipsis.
* Remove padding from ellipsis width calculation.  Only add padding in when
looking at a key where a value also exists.  No need to save space for an
ellipsis and padding otherwise.

bwmilby commented:
@livecodeali This is the code that we discussed at the conference.

bwmilby committed:
[Tree View Widget] Add enhancement note

bwmilby committed:
[Tree View Widget] Add bug fix notes

bwmilby committed:
[Tree View Widget] save `show border` property

bwmilby committed:
[Tree View Widget] Add `showValues` property

livecodeali commented on extensions/widgets/treeview/treeview.lcb:
```diff
@@ -662,6 +662,13 @@ private handler elementIsProxyForMoreData(in pElement as Array) returns Boolean
 
 public handler OnPaint() returns nothing
+	variable tRowHeight as Integer
+	put the size of my font + 8 into tRowHeight
```
Could you make this a `constant kRowTextPadding is 4` and then `the size of my font + 2 * kRowTextPadding into tRowHeight` - i suppose it's something that might become a property itself at some stage

bwmilby replied:
Added in 10/3 commit

livecodeali commented on extensions/widgets/treeview/treeview.lcb:
```diff
@@ -1285,6 +1328,13 @@ private handler updateSeparator() returns nothing
 			variable tKeyDisplay as String
 			put tElement["key"] into tKeyDisplay
+			if mCharsToTrimFromKey > 0 then
+				if the number of chars in tKeyDisplay < mCharsToTrimFromKey then
```
You might as well make this a `<=` as it will be slightly more efficient to replace the value with the empty string than delete all its chars

bwmilby replied:
Changed in 10/3 commit

livecodeali commented on extensions/widgets/treeview/treeview.lcb:
```diff
@@ -2214,9 +2264,16 @@ private handler setSortOrder(in pType as String) returns nothing
 
+private handler setCharsToTrimFromKey(in pChars as Integer) returns nothing
+	if pChars < 0 then
+		throw "Leading Characters to Trim From Key must be >= 0"
```
perhaps `number of leading characters to trim from key must be non-negative`? I think the convention here is the we do not capitalise error messages used with `throw` (although there are some in this file that are initially capitalised)

bwmilby replied:
Fixed in 10/3 commit (also made other throw's lower case)

livecodeali commented on extensions/widgets/treeview/treeview.lcb:
```diff
@@ -535,8 +535,9 @@ Parameters:
-the widget.  A value of 0 will use the default calculated height.
+Use the <textHeight> property to set a custom text (row) height for
+the widget.  The default value is 0 which will use the calculated height
+based on the currently selected font and size.
```
There's something up with the commit here - I guess you maybe put the bug fix and test changes into a different one?

bwmilby replied:
The bug never made it to a release version.  I had split off to a new branch for these updates and discovered/fixed the bug.  I went back to the original PR and fixed it there before it was merged in.  I probably should should have done it as a separate commit on this branch, but had forgotten about it when I was doing my cherry pick.

bwmilby replied:
The bug was fixed in PR #6994 prior to 9.5 release

livecodeali commented:
@bwmilby This looks great - there's just a few minor things plus the issue with the missing bugfix / test from f2facb5a2518371e9e3b37924d605f5a133f4a08

bwmilby committed:
[Tree View Widget] Row Padding Constant and Throw Message updates

added `kRowTextPadding` constant
updated all new `throw` messages to be lower case
changed trim error message

bwmilby commented:
PR #6994 is where the missing code is from https://github.com/livecode/livecode/commit/f2facb5a2518371e9e3b37924d605f5a133f4a08

bwmilby commented:
I think this would be good to be merged into 9.6.0 since it is a mobile targeted enhancement.  I'm not sure what else I can to for this to satisfy the "change requested" flag though.

livecodeali commented:
@bwmilby Sorry about that - was a bit confused about f2facb5 - I think it just has an incorrect commit message then? 
```
[Tree View Widget] Clear hilitedElement bug fix 
If no element was selected, setting the hilitedElement to empty would
throw an error.  Added an additional test to ensure that the widget did
not attempt to unselect and empty path.
```
The content of the commit seems to be a fix to the `textHeight` docs

livecodeali commented:
@bwmilby Actually ignore that - I think I will just squash merge the PR as it's all just in one file

livecodeali commented:
@livecode-vulcan review ok e7e96e9

livecode-vulcan commented:
💙 review by @livecodeali ok e7e96e9b518bd933b260a6fc8e926a576d5b5919

livecode-vulcan commented:
😞 test failure e7e96e9b518bd933b260a6fc8e926a576d5b5919

* try-community-armv6-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-sdk26_ndk16r15/builds/682)
* try-community-armv7-android-ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv7-android-ndk16r15/builds/91)
* try-community-arm64-android-ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-arm64-android-ndk16r15/builds/91)
* try-community-x86-android-ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86-android-ndk16r15/builds/91)
* try-community-x86_64-android-ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86_64-android-ndk16r15/builds/91)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/2208)
* try-community-universal-ios-iphoneos13.1: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos13.1/builds/42)
* try-community-universal-ios-iphonesimulator13.1: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator13.1/builds/42)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/2237)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/2224)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/2223)
* try-community-x86-win32: [failure](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/2257)
* try-community-x86_64-win32: [failure](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/2197)

livecodeali commented:
@bwmilby The build is complaining about the lack of docs for `scroll` because you've sort of combined the `viewTopPosition` and `scroll` properties - is there a need for the `viewTopPosition` synonym?

bwmilby commented:
Strictly there isn't a need for the synonym.  Initially I think I just exported the internally used variable name to make is accessible and added the synonym to be consistent with other controls.  I can do a commit to remove it.

livecodeali commented:
Cool yeah I think it should be fine just to use `scroll` (and therefore also `vScroll`)

bwmilby committed:
[Tree View Widget] Remove viewTopPosition property (synonym)

Remove the `scroll` synonym (`viewTopPosition`)

livecodeali commented:
@livecode-vulcan review ok 5f0c31f

livecode-vulcan commented:
💙 review by @livecodeali ok 5f0c31fcf978220fc911b849a002e9a059ef87cd

livecode-vulcan commented:
😎 test success 5f0c31fcf978220fc911b849a002e9a059ef87cd

* try-community-armv6-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-sdk26_ndk16r15/builds/683)
* try-community-armv7-android-ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv7-android-ndk16r15/builds/92)
* try-community-arm64-android-ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-arm64-android-ndk16r15/builds/92)
* try-community-x86-android-ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86-android-ndk16r15/builds/92)
* try-community-x86_64-android-ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86_64-android-ndk16r15/builds/92)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/2209)
* try-community-universal-ios-iphoneos13.1: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos13.1/builds/43)
* try-community-universal-ios-iphonesimulator13.1: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator13.1/builds/43)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/2238)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/2225)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/2224)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/2258)
* try-community-x86_64-win32: [success](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/2198)
